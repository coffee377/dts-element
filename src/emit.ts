import * as ts from 'typescript';
import {ElementKind} from './constants';
import {IElement} from './element';
import {emit_triple_slash_reference, ITripleSlashReference} from './others/triple-slash-reference';
import {transform} from './transform';

export const emit = (element: IElement<any>, options?: ts.PrinterOptions) => {
  switch (element.kind) {
    case ElementKind.TripleSlashReference:
      return emit_triple_slash_reference(element as ITripleSlashReference);
    default:
      const ts_node = transform(element);
      const source_file = ts.createSourceFile('', '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
      const printer = ts.createPrinter(options);
      return printer.printNode(ts.EmitHint.Unspecified, ts_node, source_file);
  }
};
