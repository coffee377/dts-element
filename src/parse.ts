import * as ts from 'typescript';
import {IElement} from './element';
import {create_top_level_element, ITopLevelElement} from './others/top-level-element';
import {create_triple_slash_reference} from './others/triple-slash-reference';

export const parse_native = (node: ts.Node): IElement<any> => {
  switch (node.kind) {
    case ts.SyntaxKind.SourceFile:
      const source_file = node as ts.SourceFile;
      return create_top_level_element({
        members: [
          ...source_file.referencedFiles.map(
            file => create_triple_slash_reference({
              path: file.fileName,
            }),
          ),
          ...source_file.typeReferenceDirectives.map(
            file => create_triple_slash_reference({
              types: file.fileName,
            }),
          ),
          ...source_file.statements.map(parse_native),
        ],
      });
    default:
      throw new Error(`Unexpected ts-kind ${node.kind} ( ${ts.SyntaxKind[node.kind]} )`);
  }
};

export const parse = (code: string) =>
  parse_native(ts.createSourceFile('', code, ts.ScriptTarget.Latest, /*setParentNodes */ false)) as ITopLevelElement;
