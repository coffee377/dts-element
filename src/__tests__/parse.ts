import * as ts from 'typescript';
import {emit} from '../emit';
import {parse, parse_native} from '../parse';

const code = `
/// <reference path="some-path" />
/// <reference types="some-types" />
declare var variable_declaration_var;
declare let variable_declaration_let;
declare const variable_declaration_const;
`;

it('should return correctly', () => {
  expect(emit(parse(code))).toMatchSnapshot();
});

it('should throw error with unexpected kind', () => {
  expect(() => parse_native(ts.createToken(-1))).toThrowError();
});
