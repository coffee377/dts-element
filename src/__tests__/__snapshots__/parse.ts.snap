// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correctly 1`] = `
"/// <reference path=\\"some-path\\" />
/// <reference types=\\"some-types\\" />
declare var v_var: any;
declare let v_let: any;
declare const v_const: any;
declare const v_any: any;
declare const v_boolean: boolean;
declare const v_never: never;
declare const v_null: null;
declare const v_number: number;
declare const v_object: object;
declare const v_string: string;
declare const v_symbol: symbol;
declare const v_this: this;
declare const v_undefined: undefined;
declare const v_void: void;
declare const v_array: any[];
declare const v_constructor: new () => any;
declare const v_function: () => any;
declare const v_function_predicate: () => this is string;
declare const v_general: X.Y.Z<string>;
declare const v_intersection: string & number;
declare const v_keyof: keyof X;
declare const v_literal_true: true;
declare const v_literal_false: false;
declare const v_literal_number: 123;
declare const v_literal_string: \\"str\\";
declare const v_mapped: {
    [T in keyof X]: any;
};
declare const v_mapped_readonly: {
    readonly [T in keyof X]: any;
};
declare const v_mapped_optional: {
    [T in keyof X]?: any;
};
declare const v_object: {
    readonly a?: any;
    b?(): any;
    new (): any;
    (): any;
    readonly [index: number]: any;
};
declare const v_sub: X[any][\\"abc\\"][123];
declare const v_tuple: [string, number];
declare const v_typeof: typeof X.Y.Z;
declare const v_union: string | number;
import \\"path/to/somewhere\\";
import import_default from \\"path/to/somewhere\\";
import import_default1, * as import_namespace1 from \\"path/to/somewhere\\";
import * as import_namespace2 from \\"path/to/somewhere\\";
import import_default2, { im_0 as im_1, im_2 } from \\"path/to/somewhere\\";
import { im_3 as im_4, im_5 } from \\"path/to/somewhere\\";
export default X.Y.Z;
export = A.B.C;
export { ex_0 as ex_1, ex_2 };
export * from \\"path/to/somewhere\\";
export { ex_3 as ex_4, ex_5 } from \\"path/to/somewhere\\";
export declare abstract class ABC<T, U> extends XYZ<S, V> {
    constructor();
    method?(): void;
    public method1(): void;
    protected method2(): void;
    private method3(): void;
    public abstract method4(): void;
    property?: string;
    public property1: string;
    protected property2: string;
    private property3: string;
    public abstract property4: string;
}
"
`;
